asyncy/processing/Services.py

```
@@ -438,6 +440,54 @@ def smart_insert(cls, story, line, command_conf: dict, key: str, value,
                               f'response body: {response_body}',
                               story=story, line=line)
 
+    @classmethod
+    def validate_output_properties(cls, command_conf: dict, body, story,
+                                   line):
+        """
+        Verify all properties are contained in the return body
+        """
+        output = command_conf.get('output', {})
+        body_json = {}
+        if type(body) == str:
+            body_json = ujson.loads(body)
+        elif type(body) == dict:
+            body_json = body
+        if output.get('properties') is not None and output.get(
+                'type') == 'object':
+            try:
+                cls.recurse_nested('object', output, body_json)
+                return body_json
```

The verification doesn't need to return anything as by definition it only verifies.

---

So thats how I did it earlier but then we didn't know which property is missing or whose type is not matched. Because validate would only return a True or False.

---

Throw an exception with your message inside it, instead of trying to return what was invalid.

---

Also, your recurse_method approach is a little weird. Can you make it like a normal recursive function?

---

You still don't know what's missing by returning the original JSON object. You do that by throwing exceptions. The point here was that there's no need to return the original JSON object as you're not supposed to touch it anyhow and if there's a validation error, you'll raise a validation error anyhow.